Add RepositoryVersion Filter

werwty opened the pull request:
fixes #3238
https://pulp.plan.io/issues/3238

pep8speaks commented:
Hello @werwty! Thanks for submitting the PR.



 - In the file [`pulpcore/pulpcore/app/viewsets/repository.py`](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py), following are the PEP8 issues :

> [Line 139:101](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 174:9](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py#L174): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 179:81](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py#L179): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 183:17](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py#L183): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 187:15](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py#L187): [E271](https://duckduckgo.com/?q=pep8%20E271) multiple spaces after keyword
> [Line 193:14](https://github.com/pulp/pulp/blob/15569da09852b9449973bc7df789f34906736053/pulpcore/pulpcore/app/viewsets/repository.py#L193): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator

werwty commented:
The filter can be tested with the following query: http://pulp3.dev:8000/api/v3/repositories/f6bf16e4-e419-482b-8032-942e3fa5b93b/versions/?version_range=3,4&created_0=2018-02-06 19:27&created_1=&content_in=1ac31f86-04ff-47cf-a3ff-2578dcaebeec,00ab453f-dbb0-481e-9934-cf68784989dd

daviddavis commented:
I was thinking that the ranges would be like two fields like `created_after` and `created_before`. Not sure I understand the field names `created_0` and `created_1`.

Also, I think that the content filter should accept hrefs like we do when you create a repo version with a list of content units:

https://github.com/pulp/pulp/blob/3.0-dev/pulpcore/pulpcore/app/viewsets/repository.py#L227

daviddavis commented:
It's late so apologies if I am not thinking clearly but it seems like your content filter algorithm is not removing one version number when it generates it range. In other words, if I search for a content unit that was added in version 4 and removed in version 6, will your code with search versions 4-6 or will it only search versions 4 and 5?

daviddavis commented:
Nevermind, I see that `range(4, 6)` will return `[4, 5]`. Ignore my previous comment 😄

daviddavis commented:
Is this `version_range` really a `range` filter or is it a `in` filter? I would expect `version_range=3,6` to return all versions from 3-6 whereas with `version_in=3,6` I would expect versions 3 and 6 to be returned.

werwty commented:
Thanks for the review! I updated so that instead of trying to use a fancy DateTimeFromToRangeFilter (which generated the `created_0` and `created_1` queries, that needed a custom widget to override https://django-filter.readthedocs.io/en/master/ref/widgets.html#suffixedmultiwidget ) I just separated it into two filters.

I did the same for the VersionRange filter, since a `NumberRangeFilter(BaseRangeFilter, NumberFilter)` expected both a lower and upper bound, and doing it through two filters does not.

`RepositoryVersionContentInFilter` was also updated to take content hrefs.

asmacdo commented on pulpcore/pulpcore/app/viewsets/repository.py:
```diff
@@ -128,6 +129,86 @@ class Meta:
+    def filter(self, qs, value):
+        """
+        Args:
+            qs (pulpcore.app.viewsets.repository.RepositoryVersionViewSet.queryset):
```
This type is very informative, but I think it violates the spirit of the docstring conventions. It probably ought to be:

qs (django.db.models.query.QuerySet)

asmacdo commented on pulpcore/pulpcore/app/viewsets/repository.py:
```diff
@@ -128,6 +129,86 @@ class Meta:
+        2. Get a list of version_added and version_removed where the content was
+           changed on the repository
+        3. Calculate and return the versions that the contents can be found on
+
```
No space here

werwty commented on pulpcore/pulpcore/app/viewsets/repository.py:
```diff
@@ -30,17 +32,16 @@
 
 class RepositoryFilter(filterset.FilterSet):
     name_in_list = CharInFilter(name='name', lookup_expr='in')
-    content_added_since = filters.Filter(name='last_content_added', lookup_expr='gt')
```
the last_content_added field was removed from Repository a while back

daviddavis commented:
The code looks good. I just realized though: if I filter on content_in=A,B,C and I have a version that contains only A, will it be returned?

werwty commented:
@daviddavis it would be returned. Currently this is an 'OR' filter, and would return all repositories containing (A | B | C)

daviddavis commented on pulpcore/pulpcore/app/viewsets/repository.py:
```diff
@@ -128,6 +129,87 @@ class Meta:
+        if not value:
+            return qs
+
+        contend_pks = [RepositoryVersionViewSet.get_content_unit_pk(url) for url in value]
```
[A PR just got merged that affects this](https://github.com/pulp/pulp/pull/3311). I think you just need this now:

```python
contend_pks = [GenericNamedModelViewSet.get_resource(url, Content).pk for url in value]
```

daviddavis commented:
One comment. Otherwise LGTM.

werwty commented on pulpcore/pulpcore/app/viewsets/custom_filters.py:
```diff
@@ -9,3 +9,8 @@ class CharInFilter(filters.BaseInFilter, filters.CharFilter):
+    pass
+
+
+class NumberRangeFilter(filters.BaseInFilter, filters.NumberFilter):
```
I updated this to use 2 NumberFilters:
```
version_min = filters.NumberFilter(name='number', lookup_expr='gte')
version_max = filters.NumberFilter(name='number', lookup_expr='lte')```

Mainly so we don't get `version_0=&version_1=` in the field name
